support/cas-server-support-oauth/src/test/java/org/apereo/cas/support/oauth/authenticator/OAuth20ClientIdClientSecretAuthenticatorTests.java

```diff
@@ -105,4 +114,42 @@ public void verifyAuthenticationWithAttributesMapping() {
         assertNotNull(credentials.getUserProfile().getAttribute("eduPersonAffiliation"));
         assertNull(credentials.getUserProfile().getAttribute("groupMembership"));
     }
+
+
+    @Test
+    public void verifyAuthenticationWithoutResolvedPrincipal() {
+
+        val id = "serviceWithAttributesMapping";
+
+        val nullPrincipalAuthenticator = new OAuth20ClientIdClientSecretAuthenticator(servicesManager,
```

Comment:
Appreciate the test. A few comments:

1. Could you obtain the authenticator from the application context rather than creating it manually? 
2. For the resolver, please use a mock.
3. Also try to remove the all the extra empty lines around the test method.

Reply:
Thanks for your feedback. I have reformatted the file and the class is now using the @Autowired authenticator. However I don't know how to replace the defaultPrincipalResolver with my mock in a good way. Do you have any suggestions?

Reply:
Sure. This is slightly complicated. What you have to do is:

- Create a TestConfiguration inner class and inside it, create a bean for the `defaultPrincipalResolver` and returns a mock and resolves a null principal.

- Then you break the test class into two separate inner nested test classes. You should move the existing tests for this class into this nested class. 
- For the second class, you include the TestConfiguration class as an imported test configuration, and you add your test there for the null principal.

Example for nested classes: `AmazonSecurityTokenServiceEndpointTests`


What this means is,

1. First category of tests run in the first class as usual
2. Second category of tests use your test configuration and your `defaultPrincipalResolver`, without affecting other tests in the same class

Reply:
If you look around the codebase for `@Nested` you will find other examples that might in fact be better.

Reply:
There you go!